Core/Scripts: Razorscale Rewrite

Keader opened the pull request:
**Changes proposed:**

-  Fixed Harpoons behavior
-  Implemented Engineers behavior
- Added Razorscale initial path
- Added missing spells
- Fixed Spell timers
- Added missing spellscripts
- Removed old hacks
- Added DespawnAtEvade method
- Updated codestyle
- Added missing texts
- Encounter fully rewritten

**Target branch(es):** 3.3.5

**Issues addressed:** Closes #13056, #19038

**Tests performed:** Builded and tested in game

**Known issues and TODO list:**

- [x] After wipe in perma_ground phase, he lost the "fly" animation and started walking in the air. 
- [x] Missing a cleanup xD
- [ ] For this PR to work, you need to change instance visibility to 400y.  Maybe #18067 will bring a better solution.

Many thanks to the good friends who helped me in this fight. ❤️

Keader committed:
Core/Scripts: Razorscale Rewrite

ccrs commented:
````quote
I tried leaving it as close to Blizzard as possible, but problems with the motion system do not allow the behavior to be fully copied
````
describe them

Keader commented:
I cant set speed for only 1 movement like Sniff show.

I cant reproduce a movement with this flags:

> (MovementSpline) Flags: Flying, Catmullrom, Walkmode, Parabolic (35654144)
> (MovementMonsterSpline) (MovementSpline) Flags: Flying, Walkmode (2097664)

tkrokli commented on src/server/scripts/Northrend/Ulduar/Ulduar/boss_razorscale.cpp:
```diff
@@ -15,1179 +15,1881 @@
+        }
 
-            void JustDied(Unit* /*killer*/) override
+        void SpellHit(Unit* caster, SpellInfo const* spell) override
```
```
src/server/scripts/Northrend/Ulduar/Ulduar/boss_razorscale.cpp:503:29: fatal error: 
      unused parameter 'caster' [-Wunused-parameter]
        void SpellHit(Unit* caster, SpellInfo const* spell) override
                            ^
```

Keader committed:
Implemented WorldState change and fix travis

Keader committed:
Cleanup

tkrokli commented on sql/updates/world/3.3.5/2021_80_98_12_world.sql:
```diff
@@ -0,0 +1,386 @@
+(33186,0,3,33282,606.2806,-137.2628,391.6004,4.537856,8,0); -- Razorscale Harpoon Fire State
+
+DELETE FROM `gossip_menu_option` WHERE `menu_id`=10314;
+INSERT INTO `gossip_menu_option` (`menu_id`, `id`, `option_icon`, `option_text`, `OptionBroadcastTextID`, `option_id`, `npc_option_npcflag`, `action_menu_id`, `action_poi_id`, `box_coded`, `box_money`, `box_text`, `BoxBroadcastTextID`, `VerifiedBuild`) VALUES
```
Some of the DB column names have changed recently :
```sql
DELETE FROM `gossip_menu_option` WHERE `MenuID`=10314;
INSERT INTO `gossip_menu_option` (`MenuID`,`OptionID`,`OptionIcon`,`OptionText`,`OptionBroadcastTextID`,`OptionType`,`OptionNpcFlag`,`ActionMenuID`,`ActionPoiID`,`BoxCoded`,`BoxMoney`,`BoxText`,`BoxBroadcastTextID`,`VerifiedBuild`) VALUES
```

tkrokli commented on sql/updates/world/3.3.5/2021_80_98_12_world.sql:
```diff
@@ -0,0 +1,386 @@
+(10314, 0, 0, 'We are ready to help!', 33353, 1, 1, 0, 0, 0, 0, '', 0, 0);
+
+DELETE FROM `creature_text` WHERE `entry`=33210;
+INSERT INTO `creature_text` (`entry`,`groupid`,`id`,`text`,`type`,`language`,`probability`,`emote`,`duration`,`sound`,`BroadcastTextId`,`TextRange`,`comment`) VALUES
```
Here also :
```sql
DELETE FROM `creature_text` WHERE `CreatureID`=33210;
INSERT INTO `creature_text` (`CreatureID`,`GroupID`,`ID`,`Text`,`Type`,`Language`,`Probability`,`Emote`,`Duration`,`Sound`,`BroadcastTextId`,`TextRange`,`comment`) VALUES
```

Keader replied:
why they changed it ? :(

tkrokli replied:
- See commit https://github.com/TrinityCore/TrinityCore/commit/d17a45ee7e348d6fc9146dd81e0457316eef1a62 (2017-05-04 `sql/updates/world/3.3.5/2017_05_04_00_world.sql`) for `creature_text` changes.
- See commit https://github.com/TrinityCore/TrinityCore/commit/5ff8601d36a14e95f9c227db1b85253ff36e551e (2017-05-04 `sql/updates/world/3.3.5/2017_05_04_02_world.sql` for `gossip_menu` and `gossip_menu_option` changes.

codeman8214 commented:
I don't think that is correct because of the state it is in

tkrokli commented:
OK, we may have to wait and see how travis-ci compiles this time. The previous compile included the SQL files, but for some reason they don't show up as part of the compile on my screen after the latest update.

Keader commented:
@codeman8214 what ?!

@tkrokli you are a angel <3

tkrokli commented:
`[ ... ] The previous compile included the SQL files, but for some reason they don't show up as part of the compile on my screen after the latest update.`

https://www.traviscistatus.com/incidents/bnt2wtxpgs39
- [**Delays for `sudo: required` builds on both .com and .org**
Incident Report for Travis CI](https://www.traviscistatus.com/incidents/bnt2wtxpgs39)

Keader committed:
Travis and Tkrokli feedback

Keader committed:
Fix some sql errors

codeman8214 commented:
Looks good now

ariel- commented:
> I can't reproduce a movement with these flags:
``` (MovementSpline) Flags: Flying, Catmullrom, Walkmode, Parabolic (35654144)```

The problem is here:

https://github.com/TrinityCore/TrinityCore/blob/3.3.5/src/server/game/Movement/Spline/MoveSplineFlag.h#L101

Setting Flying will unset Catmullrom and viceversa.

Shauren commented:
@ariel- @Keader that sniff parse to me looks like a wod/legion wpp format with the substructures. If that is the case then you need to remember that spline flag combinations are different between 3.3.5 and cataclysm+ (starting with cata they are mostly the same until latest version). In 3.3.5 flying & catmullrom are never together but in legion they can be

Keader committed:
Sirikfoll feedbacks and some improvements

Keader commented:
@ccrs why we dont have some function that allow pass custom MoveSpline, and pass flags to MoveSpline?
e.g.
```cpp
Movement::MoveSplineInit init(me);
init.SetVelocity(21.f);
init.MoveTo(RazorscalePoints[0]);
init.SetFlags(Movement::MoveSplineFlag::Flying | Movement::MoveSplineFlag::Walkmode);
me->GetMotionMaster()->MoveCustomSpline(POINT_COMBAT, init);
```

Shauren commented:
Walkmode name is wrong and has already been renamed in master (applies to 3.3.5 too) - this flag means CanSwim
Flying is set automatically be spline when DisableGravity or CanFly are enabled

ccrs commented:
what sharuen said + you can always send your spline and use a controlled movement to keep it like movepoint with charge id (check gunship battle)

Riztazz commented:
https://github.com/TrinityCore/TrinityCore/blob/3.3.5/src/server/game/Entities/Creature/Creature.cpp#L2913

Add that flag and handle movement within script, this part of code is probably overwriting your stuff

ccrs committed:
testing canfly flag

Keader commented:
yay, something is wrong (idk if is core issue, or i fuckedup something)
but using:

```cpp
DummyEntryCheckPredicate pred;
summons.DoAction(ACTION_START_FIGHT, pred);
```

This action ACTION_START_FIGHT is not getted for any creature in summon list
i put comments in ```DoAction``` method to all summon and is not called

ariel- commented:
Because SummonList::DoAction has a third optional parameter, max list size, defaults 0 and it resizes the list to 0 elements basically never triggering the event for any summons :c

ariel- commented:
Rebase on 8188511759c6613d33111ab59f4af3c441d01c8b to get rid of the DoAction bug

ccrs committed:
little cleaning

ccrs committed:
misc

erase delayed orientation change, revert POINT_START_WAYPOINT, move event update call, erase unneded spelldifficulty

ccrs commented:
we can erase the "movement sytem is bugged" cloud now 💃

Keader committed:
Devouring Flame issue and Path issue fixed

Keader committed:
Typo

ccrs commented on sql/updates/world/3.3.5/2021_80_98_12_world.sql:
```diff
@@ -0,0 +1,381 @@
+UPDATE `creature_text` SET `comment`='Expedition Engineer SAY_START_REPAIR' WHERE `CreatureID`='33287' AND `GroupID`=1;
+UPDATE `creature_text` SET `comment`='Expedition Engineer SAY_REBUILD_TURRETS' WHERE `CreatureID`='33287' AND `GroupID`=2;
+
+UPDATE `creature_template` SET `InhabitType`=5 WHERE `entry`IN (33186, 33724); -- Razorscale
```
you set canfly as default and use disablegravity on the script.
first of all hack, secondly like I told you you have to find which of them is the correct. You cant use both, only one active.

Keader replied:
i forgot to remove it xd
btw...

https://github.com/TrinityCore/TrinityCore/blob/3.3.5/src/server/scripts/Northrend/ChamberOfAspects/RubySanctum/boss_saviana_ragefire.cpp#L161-L162

https://github.com/TrinityCore/TrinityCore/blob/3.3.5/src/server/scripts/Northrend/IcecrownCitadel/boss_sindragosa.cpp#L318-L319

Should have more dragon bosses using both

ccrs replied:
hacks in both cases and in any other case.

ccrs commented:
test it now with a clean database, or with the original inhabit value

sirikfoll commented:
InhabitType=4 is the correct, and aparently there's not even need to use those me->SetDisableGravity(false/true), it already works well leaving it for the core to handle.

@ccrs since you seem to understand about those movement problems, do you have any inputs on this issue? "After a wipe while boss is in perma_ground phase, he will respawn with bugged fly animation(walking in the air)."

Keader commented:
@ariel- can you review includes after cleanup port in razorscale.cpp please?

ariel- commented on src/server/scripts/Northrend/Ulduar/Ulduar/boss_razorscale.cpp:
```diff
@@ -15,15 +15,16 @@
+#include "MoveSplineInit.h"
 #include "ObjectAccessor.h"
+#include "PassiveAI.h"
+#include "PathGenerator.h"
```
Can't see what you need PathGenerator for

Keader replied:
without it, i got error in:

```cpp
uint32 const pathSize = 11;
G3D::Vector3 const RazorscalePath[pathSize] =
```
"Error (active)		incomplete type is not allowed	scripts_northrend	c:\Users\Bradley\Documents\Trinity\src\server\scripts\Northrend\Ulduar\Ulduar\boss_razorscale.cpp	279	
"

Keader replied:
Should be changed to #include <G3D/Vector3.h> right?
And what the rule to put this includes? alphabetical order?

ariel- replied:
Yes and yes

ariel- commented on src/server/scripts/Northrend/Ulduar/Ulduar/boss_razorscale.cpp:
```diff
@@ -15,1185 +15,1902 @@
-const Position RazorFlight = { 588.050f, -251.191f, 470.536f, 1.498f };
-const Position RazorGround = { 586.966f, -175.534f, GROUND_Z, 4.682f };
-const Position PosEngSpawn = { 591.951f, -95.9680f, GROUND_Z, 0.000f };
+const Position RazorFlightPosition       = { 585.3610f, -173.5592f, 456.8430f, 1.526665f };
```
Position const (codestyle and consistency with rest of consts)

Keader commented:
i'm thinking about merge it...
Using wrong spawn point (currently spawn point) and not using path part, until have a way to change object distance view...

tkrokli commented:
Oh, right. I almost thought this was merged already.

ccrs commented:
also the bug that made the fire chase players is solved, u can remove the ai

Keader commented:
ok ccrs, i will do it tomorrow morning, ty.

Keader committed:
Last changes

Keader committed:
sql rename
